homeassistant/components/sensor/github.py

```diff
@@ -0,0 +1,294 @@
+"""
+Support for GitHub.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.github/
+"""
+from datetime import timedelta
+import logging
+import json
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_ACCESS_TOKEN, CONF_USERNAME, CONF_PASSWORD, CONF_PATH, CONF_NAME,
+    CONF_SCAN_INTERVAL, ATTR_NAME)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+REQUIREMENTS = ['PyGithub==1.43.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_GROUP_AUTH = 'authentication'
+CONF_REPOS = 'repositories'
+CONF_SERVER_URL = 'server_url'
+
+ATTR_PATH = 'path'
+ATTR_STARGAZERS = 'stargazers'
+ATTR_TOPICS = 'topics'
+ATTR_BRANCHES = 'branches'
+ATTR_OPEN_ISSUES = 'open_issues'
+ATTR_OPEN_PULL_REQUESTS = 'open_pull_requests'
+ATTR_LAST_COMMIT = 'last_commit'
+ATTR_LATEST_RELEASE = 'latest_release'
+
+DEFAULT_NAME = 'GitHub'
+
+SCAN_INTERVAL = timedelta(seconds=300)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Exclusive(CONF_ACCESS_TOKEN, CONF_GROUP_AUTH): cv.string,
+    vol.Exclusive(CONF_USERNAME, CONF_GROUP_AUTH): cv.string,
+    vol.Optional(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_SERVER_URL): cv.url,
+    vol.Required(CONF_REPOS): cv.ensure_list
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the GitHub sensor platform."""
+    if config.get(CONF_ACCESS_TOKEN) is None \
+            and config.get(CONF_USERNAME) is None:
+        # Error if there is no authentication set
+        _LOGGER.error("You have not set an access_token or a username")
+        return
+
+    interval = config.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL)
+    sensors = []
+    for repository in config[CONF_REPOS]:
+        sensors.append(GitHubSensor(GitHubData(
+            interval=interval,
+            repository=repository,
+            access_token=config.get(CONF_ACCESS_TOKEN),
+            username=config.get(CONF_USERNAME),
+            password=config.get(CONF_PASSWORD),
+            server_url=config.get(CONF_SERVER_URL)
+        )))
```

Comment:
The platform setup should fail if the credentials are wrong. It might be that the repo is not available at the moment you start HA but from invalid credential can the sensor not recover.